File: src/services/transcription.js
import fs from 'node:fs';
import OpenAI from 'openai';
import { v4 as uuidv4 } from 'uuid';
import logger from '../logger.js';
import { calculateDurationMs } from '../utils/index.js';
import {
  markTranscribed,
  markFailed,
  markInProgress,
  markSkipped
} from './manifest.js';

const openai = new OpenAI();

// Error types that indicate we should not retry
const PERMANENT_ERRORS = [
  'invalid_api_key',
  'audio_format'
];

// Error types that indicate API quota/rate issues (recoverable later)
const QUOTA_ERRORS = [
  'insufficient_quota',
  'rate_limit_exceeded'
];

/**
 * Check if error is a quota/rate limit error
 */
function isQuotaError(err) {
  if (err.status === 429) return true;
  if (err.code === 'insufficient_quota') return true;
  if (QUOTA_ERRORS.some(e => err.message?.includes(e))) return true;
  return false;
}

/**
 * Check if error is permanent (no point retrying)
 */
function isPermanentError(err) {
  if (err.status === 401) return true;
  if (err.status === 400 && err.message?.includes('audio_format')) return true;
  if (PERMANENT_ERRORS.some(e => err.message?.includes(e))) return true;
  return false;
}

/**
 * Get human-readable error message
 */
function getErrorMessage(err) {
  if (err.status === 401) {
    return 'Invalid API key';
  } else if (err.status === 429) {
    return 'Rate limit exceeded or insufficient credits';
  } else if (err.code === 'insufficient_quota') {
    return 'Quota exceeded - please add credits';
  } else if (err.status === 400) {
    return 'Bad request - check audio format';
  } else if (err.status === 413) {
    return 'File too large (max 25MB)';
  } else if (err.message?.includes('Connection') || err.message?.includes('timeout')) {
    return 'Connection error';
  } else if (err.message?.includes('audio_format')) {
    return 'Unsupported audio format';
  }
  return err.message || 'Unknown error';
}

/**
 * Transcribe an audio file using OpenAI Whisper
 * @param {string} wavPath - Path to the WAV file
 * @param {string} sessionId - Session identifier for logging
 * @param {string} userId - User ID for logging
 * @param {string} entryId - Manifest entry ID (optional)
 * @returns {Promise<{text: string, success: boolean, error: string|null}>}
 */
export async function transcribeAudio(wavPath, sessionId, userId, entryId = null) {
  if (!wavPath) {
    return { text: '', success: false, error: 'No audio path provided' };
  }

  const transcribeEventId = uuidv4();
  const startTime = Date.now();

  // Mark as in progress in manifest
  if (entryId) {
    markInProgress(entryId);
  }

  try {
    // Check if file exists and has content
    if (!fs.existsSync(wavPath)) {
      const error = 'Audio file not found';
      console.error(`❌ ${error}: ${wavPath}`);
      if (entryId) {
        markFailed(entryId, error);
      }
      return { text: '', success: false, error };
    }

    const stats = fs.statSync(wavPath);
    if (stats.size < 1000) {
      const error = `File too small: ${stats.size} bytes`;
      console.log(`⚠️ Skipping transcription - ${error}`);
      if (entryId) {
        markSkipped(entryId, error);
      }
      return { text: '', success: false, error };
    }

    const resp = await openai.audio.transcriptions.create({
      file: fs.createReadStream(wavPath),
      model: 'whisper-1',
      language: undefined, // Auto-detect language (supports German, English, etc.)
    });

    const text = resp.text?.trim() || '';

    if (text) {
      logger.info("Transcription successful", {
        extra: {
          footprint: null,
          batch_uuid: sessionId,
          user_id: userId,
          event_id: transcribeEventId,
          action: "audio_transcription",
          event: "complete",
          duration_ms: calculateDurationMs(startTime)
        }
      });

      // Update manifest with successful transcription
      if (entryId) {
        markTranscribed(entryId, text);
      }
      return { text, success: true, error: null };
    }

    // Whisper succeeded but returned empty text (couldn't understand audio)
    const emptyReason = 'No speech detected or audio unclear';
    console.log(`⚠️ Whisper returned empty text for ${wavPath.split('/').pop()}`);
    if (entryId) {
      markSkipped(entryId, emptyReason);
    }
    return { text: '', success: false, error: emptyReason };

  } catch (err) {
    const errorMessage = getErrorMessage(err);
    const isQuota = isQuotaError(err);
    const isPermanent = isPermanentError(err);

    logger.error("Audio transcription failed", {
      extra: {
        footprint: null,
        batch_uuid: sessionId,
        user_id: userId,
        event_id: transcribeEventId,
        action: "audio_transcription",
        event: "error",
        duration_ms: calculateDurationMs(startTime),
        error_type: err.constructor.name,
        error_message: err.message,
        error_code: err.code,
        error_status: err.status,
        is_quota_error: isQuota,
        is_permanent: isPermanent
      }
    });

    console.error(`❌ Whisper failed: ${errorMessage}`);

    // Update manifest with failure
    if (entryId) {
      if (isPermanent) {
        markSkipped(entryId, `Permanent error: ${errorMessage}`);
      } else {
        markFailed(entryId, errorMessage);
      }
    }

    // Don't throw for quota errors - we want to save the audio and retry later
    if (isQuota) {
      console.log('💰 Audio saved for later retry when API credits are available');
      return { text: '', success: false, error: errorMessage, isQuotaError: true };
    }

    // For other errors, return failure but don't throw
    return { text: '', success: false, error: errorMessage };
  }
}

/**
 * Transcribe with retry logic (for recovery service)
 * @param {string} wavPath - Path to the WAV file
 * @param {string} sessionId - Session identifier
 * @param {string} userId - User ID
 * @param {string} entryId - Manifest entry ID
 * @param {number} maxRetries - Maximum retry attempts
 * @returns {Promise<{text: string, success: boolean, error: string|null}>}
 */
export async function transcribeWithRetry(wavPath, sessionId, userId, entryId, maxRetries = 3) {
  let lastError = null;

  for (let attempt = 0; attempt < maxRetries; attempt++) {
    if (attempt > 0) {
      const delay = Math.min(15000, 1000 * Math.pow(2, attempt));
      console.log(`🔄 Retry ${attempt + 1}/${maxRetries} in ${delay}ms...`);
      await new Promise(resolve => setTimeout(resolve, delay));
    }

    const result = await transcribeAudio(wavPath, sessionId, userId, entryId);

    if (result.success) {
      return result;
    }

    lastError = result.error;

    // Don't retry quota errors or permanent errors
    if (result.isQuotaError) {
      console.log('💰 Stopping retries due to quota error - will retry on next startup');
      return result;
    }
  }

  return { text: '', success: false, error: lastError };
}

/**
 * Validate OpenAI API key at startup
 * @param {string} startupEventId - Event ID for logging
 * @returns {Promise<boolean>} Whether the key is valid
 */
export async function validateOpenAIKey(startupEventId) {
  const validationEventId = uuidv4();
  const startTime = Date.now();

  logger.debug("Validating OpenAI API key", {
    extra: {
      footprint: null,
      batch_uuid: startupEventId,
      user_id: null,
      event_id: validationEventId,
      action: "openai_validation",
      event: "start"
    }
  });

  try {
    await openai.models.list();

    logger.info("OpenAI API key validation successful", {
      extra: {
        footprint: null,
        batch_uuid: startupEventId,
        user_id: null,
        event_id: validationEventId,
        action: "openai_validation",
        event: "complete",
        duration_ms: calculateDurationMs(startTime)
      }
    });

    return true;
  } catch (error) {
    logger.error("OpenAI API key validation failed", {
      extra: {
        footprint: null,
        batch_uuid: startupEventId,
        user_id: null,
        event_id: validationEventId,
        action: "openai_validation",
        event: "error",
        duration_ms: calculateDurationMs(startTime),
        error_type: error.constructor.name,
        error_message: error.message,
        error_code: error.code,
        error_status: error.status
      }
    });

    if (error.status === 401) {
      logger.error("Invalid OpenAI API key - check your OPENAI_API_KEY environment variable");
    } else if (error.status === 429) {
      logger.error("OpenAI API rate limit exceeded or insufficient credits");
    } else if (error.code === 'insufficient_quota') {
      logger.error("OpenAI API quota exceeded - please add credits to your account");
    }

    return false;
  }
}

File: summarise_bot/transcription.py
from __future__ import annotations

import array
import math
import re
import warnings
from pathlib import Path
import wave

from openai import AsyncOpenAI

from .config import SETTINGS
from .logger import logger
from .manifest import mark_failed, mark_in_progress, mark_skipped, mark_transcribed


client = AsyncOpenAI(api_key=SETTINGS.openai_api_key)

# OpenAI rejects clips under 0.1s ("audio_too_short"). Skip anything below this locally so
# empty/blip segments never hit the API (and never get retried forever by the recovery loop).
MIN_AUDIO_SECONDS = 0.15

# audioop is a C accelerator only; it is deprecated in 3.12 and REMOVED in 3.13, so it is
# optional and the pure-stdlib fallback below is used when it is gone.
try:
    with warnings.catch_warnings():
        warnings.simplefilter("ignore", DeprecationWarning)
        import audioop as _audioop  # type: ignore[import-not-found]
except ImportError:  # pragma: no cover - Python 3.13+
    _audioop = None

_AUDIO_GATE_FRAME_MS = 20

# Whisper rewrites silence and short non-speech bursts into caption boilerplate. Measured over
# 944 real captures these four strings alone account for a third of all transcribed segments.
# Only whole-line matches are dropped, so the same words survive inside a real utterance.
# Deliberately NOT listed: "yeah.", "okay.", "so.", "thanks.", "see you." — indistinguishable
# from genuine backchannels, and they carry no report content either way.
_HALLUCINATION_EXACT = {
    "you", "thank you", "thanks", "bye", "bye bye", "goodbye", "peace",
    "mm hmm", "mmhmm", "mhm", "mm", "uh huh", "hmm", "",
    "thank you very much", "thank you so much", "thanks for watching",
    "thanks for listening", "thank you for watching", "thank you for joining us",
    "please subscribe", "subscribe to my channel",
}
_HALLUCINATION_SUBSTRINGS = (
    "ontario website",
    "thanks for watching",
    "thanks for listening",
    "see you in the next",
    "new revised standard version",
    "amara.org",
    "subtitles by",
)
_PUNCT_RE = re.compile(r"[^\w\s]", re.UNICODE)


def _normalize_for_blocklist(text: str) -> str:
    """Punctuation becomes whitespace so "Bye-bye." and "Bye bye" collapse to the same key."""
    return " ".join(_PUNCT_RE.sub(" ", text).split()).lower()


# Below this, whatever precedes a hallucinated tail is not worth keeping on its own.
_MIN_SALVAGE_CHARS = 40


def _is_hallucination(text: str) -> bool:
    """True only when the ENTIRE line is filler, so filler inside a real utterance survives."""
    return _normalize_for_blocklist(text) in _HALLUCINATION_EXACT


def strip_hallucinated_tail(text: str) -> str:
    """Trim caption boilerplate appended to the end of a real utterance.

    Whisper often finishes a genuine segment with "Thanks for watching." or a subtitle credit.
    Dropping the whole line for that would lose real content, so cut at the marker and keep the
    prefix when enough of it remains; otherwise the line was filler all along.
    """
    lowered = text.lower()
    cut = min(
        (lowered.find(marker) for marker in _HALLUCINATION_SUBSTRINGS if marker in lowered),
        default=-1,
    )
    if cut < 0:
        return text
    prefix = text[:cut].strip(" \t-—–,;:")
    return prefix if len(prefix) >= _MIN_SALVAGE_CHARS else ""


def _read_wav_pcm(path: Path) -> tuple[bytes, int, int, int]:
    """(raw_frames, sample_rate, channels, sample_width); zeroed on any read failure."""
    try:
        with wave.open(str(path), "rb") as wav:
            return (
                wav.readframes(wav.getnframes()),
                wav.getframerate(),
                wav.getnchannels(),
                wav.getsampwidth(),
            )
    except (wave.Error, OSError, EOFError, MemoryError):
        return b"", 0, 0, 0


def _measure_audio(path: Path) -> tuple[float, float]:
    """Return (total_rms, voiced_fraction), or (-1.0, -1.0) when it cannot be measured.

    Roughly 98% of these captures are dual-mono, and the pooled/mono RMS ratio is 1.0000 at
    p5/p50/p95, so RMS is taken over the interleaved stream with no downmix step.
    Unmeasurable audio fails OPEN — it still goes to the API rather than being dropped blind.
    """
    raw, rate, channels, width = _read_wav_pcm(path)
    if width != 2 or not raw or not rate or not channels:
        return -1.0, -1.0

    frame_samples = max(1, int(rate * _AUDIO_GATE_FRAME_MS / 1000)) * channels
    frame_bytes = frame_samples * 2
    floor = SETTINGS.whisper_frame_rms

    if _audioop is not None:
        total = float(_audioop.rms(raw, 2))
        frames = voiced = 0
        for offset in range(0, len(raw) - frame_bytes + 1, frame_bytes):
            frames += 1
            if _audioop.rms(raw[offset : offset + frame_bytes], 2) > floor:
                voiced += 1
    else:  # pragma: no cover - exercised on Python 3.13+
        samples = array.array("h")
        samples.frombytes(raw[: len(raw) - (len(raw) % 2)])
        if sys_byteorder_is_big := (array.array("h", b"\x01\x00")[0] != 1):
            samples.byteswap()
        total = math.sqrt(sum(v * v for v in samples) / len(samples)) if samples else 0.0
        frames = voiced = 0
        for offset in range(0, len(samples) - frame_samples + 1, frame_samples):
            window = samples[offset : offset + frame_samples]
            if math.sqrt(sum(v * v for v in window) / len(window)) > floor:
                voiced += 1
            frames += 1

    return total, (voiced / frames if frames else 0.0)


def _passes_audio_gate(path: Path) -> tuple[bool, str]:
    """Block near-silent clips before they reach the API.

    Calibrated on 944 real captures joined to their transcripts: at RMS>=250 with a 10% voiced
    fraction this blocks 76% of known hallucinations and 0 of 383 substantive segments
    (real p5 RMS is 708, a 2.8x margin). It cannot catch the ~24% of hallucinations that are
    loud short bursts — a cough or a chair — which is what the logprob and blocklist layers are for.
    """
    if SETTINGS.whisper_min_rms <= 0:
        return True, ""
    rms, voiced = _measure_audio(path)
    if rms < 0:
        return True, ""  # unmeasurable: fail open
    if rms < SETTINGS.whisper_min_rms or voiced < SETTINGS.whisper_min_voiced_fraction:
        return False, f"Below speech threshold: rms={rms:.0f} voiced={voiced:.2f}"
    return True, ""


def _verbose_text(response: object) -> str:
    """Text of a verbose_json response with hallucinated segments removed."""
    segments = getattr(response, "segments", None)
    full_text = (getattr(response, "text", "") or "").strip()
    if not segments:
        return full_text
    kept: list[str] = []
    for segment in segments:
        no_speech = getattr(segment, "no_speech_prob", 0.0) or 0.0
        avg_logprob = getattr(segment, "avg_logprob", 0.0) or 0.0
        if no_speech > SETTINGS.whisper_max_no_speech_prob and avg_logprob < SETTINGS.whisper_min_avg_logprob:
            continue
        piece = (getattr(segment, "text", "") or "").strip()
        if piece:
            kept.append(piece)
    return " ".join(kept).strip()


def _wav_duration_seconds(path: Path) -> float:
    try:
        with wave.open(str(path), "rb") as wav:
            frames = wav.getnframes()
            rate = wav.getframerate()
    except (wave.Error, OSError, EOFError):
        return 0.0
    return frames / rate if rate else 0.0


def _is_quota_error(error: Exception) -> bool:
    message = str(error).lower()
    status = getattr(error, "status_code", None) or getattr(error, "status", None)
    code = getattr(error, "code", None)
    return status == 429 or code == "insufficient_quota" or "insufficient_quota" in message


def _is_permanent_error(error: Exception) -> bool:
    status = getattr(error, "status_code", None) or getattr(error, "status", None)
    message = str(error).lower()
    return (
        status == 401
        or "audio_format" in message
        or "audio_too_short" in message
        or "audio file is too short" in message
    )


async def validate_openai_key() -> bool:
    try:
        await client.models.list()
        logger.info("OpenAI API key validation successful", extra={"action": "openai_validation", "event": "complete"})
        return True
    except Exception as error:  # pragma: no cover - network exercised manually
        logger.error(
            "OpenAI API key validation failed",
            extra={"action": "openai_validation", "event": "error", "error_message": str(error)},
        )
        return False


async def close_client() -> None:
    await client.close()


async def transcribe_audio(
    wav_path: str,
    session_id: str,
    user_id: str,
    entry_id: str | None = None,
) -> dict[str, object]:
    path = Path(wav_path)
    if entry_id:
        mark_in_progress(entry_id)

    if not path.exists():
        message = "Audio file not found"
        if entry_id:
            mark_skipped(entry_id, message)  # missing file can never transcribe — don't retry forever
        return {"text": "", "success": False, "error": message, "skipped": True}

    duration_seconds = _wav_duration_seconds(path)
    if duration_seconds < MIN_AUDIO_SECONDS:
        message = f"Audio too short: {duration_seconds:.3f}s"
        if entry_id:
            mark_skipped(entry_id, message)
        return {"text": "", "success": False, "error": message, "skipped": True}

    allowed, gate_reason = _passes_audio_gate(path)
    if not allowed:
        if entry_id:
            mark_skipped(entry_id, gate_reason)
        return {"text": "", "success": False, "error": gate_reason, "skipped": True, "gated": True}

    try:
        with path.open("rb") as audio_file:
            response = await client.audio.transcriptions.create(
                file=audio_file,
                model=SETTINGS.whisper_model,
                prompt=SETTINGS.whisper_prompt,
                temperature=SETTINGS.whisper_temperature,
                response_format="verbose_json",
            )
        text = _verbose_text(response)
        if text and _is_hallucination(text):
            message = "Filtered Whisper hallucination"
            if entry_id:
                mark_skipped(entry_id, message)
            return {"text": "", "success": False, "error": message, "skipped": True, "gated": True}
        if text:
            if entry_id:
                mark_transcribed(entry_id, text)
            logger.info(
                "Transcription successful",
                extra={"action": "audio_transcription", "event": "complete", "session_id": session_id, "user_id": user_id},
            )
            return {"text": text, "success": True, "error": None}
        message = "No speech detected or audio unclear"
        if entry_id:
            mark_skipped(entry_id, message)
        return {"text": "", "success": False, "error": message, "skipped": True}
    except Exception as error:  # pragma: no cover - network exercised manually
        quota = _is_quota_error(error)
        permanent = _is_permanent_error(error)
        message = str(error)
        if entry_id:
            if permanent:
                mark_skipped(entry_id, f"Permanent error: {message}")
            else:
                mark_failed(entry_id, message)
        if permanent:
            # Not a real failure — the clip can never transcribe (e.g. too short). Skip
            # quietly so it is not logged as an error or retried by the recovery loop.
            logger.info(
                "Audio transcription skipped (permanent)",
                extra={
                    "action": "audio_transcription",
                    "event": "skipped_permanent",
                    "session_id": session_id,
                    "user_id": user_id,
                    "error_message": message,
                },
            )
        else:
            logger.error(
                "Audio transcription failed",
                extra={
                    "action": "audio_transcription",
                    "event": "error",
                    "session_id": session_id,
                    "user_id": user_id,
                    "error_message": message,
                    "is_quota_error": quota,
                },
            )
        return {"text": "", "success": False, "error": message, "isQuotaError": quota, "skipped": permanent}


async def transcribe_with_retry(
    wav_path: str,
    session_id: str,
    user_id: str,
    entry_id: str,
    max_retries: int = 3,
) -> dict[str, object]:
    last_result: dict[str, object] = {"text": "", "success": False, "error": "unknown"}
    for attempt in range(max_retries):
        if attempt:
            await __import__("asyncio").sleep(min(15, 2**attempt))
        last_result = await transcribe_audio(wav_path, session_id, user_id, entry_id)
        if last_result.get("success"):
            return last_result
        if last_result.get("isQuotaError"):
            return last_result
        if last_result.get("skipped"):
            # Permanent / too-short: retrying cannot help, so stop immediately.
            return last_result
    return last_result
